benchmarks/swe_bench/swe_bench.py

```diff
@@ -0,0 +1,374 @@
+
+        if any(error_string_search.values()):
+            explanation = f"The tests did not run correctly. Output from searching for error strings:\n\n{error_string_search}\n\nOutput from tests:\n\n{eval_output.stdout}"
+            value = 0.0
+        else:
+            test_output_parser = MAP_REPO_TO_PARSER[state.metadata["repo"]]
+            test_outputs = test_output_parser(eval_output.stdout)
+
+            pass_to_pass_results = {}
+            fail_to_pass_results = {}
+            for k, v in test_outputs.items():
+                has_passed = "PASSED" == v
+                if k in state.metadata["PASS_TO_PASS"]:
+                    pass_to_pass_results[k] = has_passed
+                else:
+                    fail_to_pass_results[k] = has_passed
+
+            # Sort both so the the false values are at the top
+            pass_to_pass_results, fail_to_pass_results = (
+                dict(sorted(pass_to_pass_results.items(), key=lambda x: x[1])),
+                dict(sorted(fail_to_pass_results.items(), key=lambda x: x[1])),
+            )
+
+            explanation = f"PASS_TO_PASS:\n\n{json.dumps(pass_to_pass_results,indent=2)}\n\nFAIL_TO_PASS:\n\n{json.dumps(fail_to_pass_results,indent=2)}\n\n"
+
+            if all(pass_to_pass_results.values()) and all(
+                fail_to_pass_results.values()
+            ):
+                value = 1.0
+            else:
+                value = 0.0
+
+        return Score(
+            value=value,
+            explanation=explanation,
+            metadata={"model_patch": agent_patch.stdout},
+        )
+
+    return scorer
+
+
+def swebench_baseline_scorer(path_to_baseline: str, name: str | None = None) -> Scorer:
+    """Given a path to a set of SWE-bench trajectories in the official format (see https://github.com/swe-bench/experiments), returns the performance of those trajectories on the subset of SWE-bench you are evaluating on. This lets you compare to baselines on arbitrary subsets of SWE-bench."""
+    baseline_name = name if name else Path(path_to_baseline).name
+
+    path_to_logs = os.path.join(path_to_baseline, "logs")
+
+    results_per_instance_id = {}
+    for result in os.listdir(path_to_logs):
+        results_path = os.path.join(path_to_logs, result, "report.json")
+        patch_path = os.path.join(path_to_logs, result, "patch.diff")
+
+        if os.path.exists(results_path) and os.path.exists(patch_path):
+            # Sometimes there is no result saved, at which point we ignore that entry
+            with open(results_path, "r") as f:
+                result_dict = json.load(f)
+                instance_id, raw_results = next(iter(result_dict.items()))
+
+                with open(patch_path, "r") as f:
+                    results_per_instance_id[instance_id] = {
+                        "resolved": raw_results["resolved"],
+                        "patch": f.read(),
+                    }
+
+    @scorer(metrics=[mean(), std()], name=baseline_name)
```

Comment:
Interested in your thoughts on the design of this. There's several nested things here, which is a little awkward I think.